### PR TITLE
Feature/usb events

### PR DESCRIPTION
**Pull request by xross**
This PR adds functionality to the test-bench to perform USB events other than simple packets.
It also makes the test-bench more closely follow OO design principles

Note, has been updated for py3

### CODE REVIEW

**Comment by xross** [comment]
@kevinyeungxmos @mbanth @DaveAtkinsonXMOS put in place most the ground work for adding other events other than packets now. All old tests now passed, taken updates from upstream/DEV/XS3 and basic conversion to python 3. Still needs some tidying but I figure you may as well take a look now.

File: tests/usb_packet.py [hunk 1]
```diff
@@ -215,63 +226,212 @@ def __init__(self, **kwargs):
...
+                    print("{0:#x}".format(item))
+
+
 
 #Tx from host i.e. xCORE Rx
 class TxPacket(UsbPacket):
 
     def __init__(self, **kwargs):
-        self.inter_pkt_gap = kwargs.pop('inter_pkt_gap', RX_RX_DELAY) #13 lowest working for single issue loopback
+        #self.inter_pkt_gap = kwargs.pop('inter_pkt_gap', RX_RX_DELAY) #13 lowest working for single issue loopback
```
Comment (DaveAtkinsonXMOS) [inline tests/usb_packet.py:347]: Can this be deleted?

File: tests/usb_packet.py [hunk 2]
```diff
@@ -68,46 +69,56 @@
...
-    data_valid_count = kwargs.pop('data_valid_count', 0)
-    packets.append(TokenPacket( 
-        inter_pkt_gap=ipg, 
-        pid=_pid,
-        address=addr, 
-        endpoint=ep,
-        data_valid_count=data_valid_count))
+#    ep = (framenumber >> 7) & 0xf
+#    address = (framenumber) & 0x7f
+#    AppendTokenPacket(packets, 0xa5, ep, ipg, address, **kwargs)
```
Comment (DaveAtkinsonXMOS) [inline tests/usb_packet.py:111]: Any reason not to delete all this commented-out code?

File: tests/usb_packet.py [hunk 3]
```diff
@@ -363,8 +521,9 @@ def __init__(self, **kwargs):
         self.crc5 = kwargs.pop('crc5', crc5)
 
         # Always override to match IFM
+        # Only required for < XS3?
         #self.data_valid_count = 4 #todo
-        self.data_valid_count = 0
+        #self.data_valid_count = 0
```
Comment (DaveAtkinsonXMOS) [inline tests/usb_packet.py:526]: Can we delete this?

File: tests/usb_transaction.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+       
+            if expectHandshake:
+                packets.append(RxHandshakePacket())
+       
+            self._packets.extend(packets)
+
+            #if resend:
+                # Add again..
+            #    self._packets.extend(packets)
+
```
Comment (DaveAtkinsonXMOS) [inline tests/usb_transaction.py:75]: Any reason not to delete this?

**Review by DaveAtkinsonXMOS — COMMENTED** [review]
Left a few comments, I don't know if the things I commented on were overlooked, or intended...

File: tests/test_bulk_rx_basic_badcrc32/src/main.xc [hunk 5]
```diff
@@ -40,11 +40,11 @@ int TestEp_Bulk(chanend c_out, chanend c_in, int epNum, chanend c_out_0)
...
-        if(i == 12)
-        unsafe{
-            g_rxDataCheck[epNum]+=12;
-            continue;   
-        }
+        //if(i == 12)
+        //unsafe{
+        //    g_rxDataCheck[epNum]+=12;
+        //    continue;   
+        //}
```
[the remark this answers is not in the document]
Comment (xross) [reply tests/test_bulk_rx_basic_badcrc32/src/main.xc:47]: Removed (Originally the test code didn't re-send the same packet. It does now)

File: tests/test_bulk_rx_traffic.py [hunk 6]
```diff
@@ -1,52 +1,36 @@
...
-    dataval = 0;
-
-    pid = PID_DATA0;
-
+    ied = 500
+   
     trafficAddress1 = 0;
     trafficAddress2 = 127;
-
     trafficEp1 = 15;
```
[the remark this answers is not in the document]
Comment (xross) [reply tests/test_bulk_rx_traffic.py]: Done

File: tests/test_bulk_rx_traffic.py [hunk 7]
```diff
@@ -55,9 +39,9 @@ def do_test(arch, clk, phy, data_valid_count, usb_speed, seed):
         trafficEp2 + trafficEp2 + 1
         if(trafficEp1 > 15):
```
[the remark this answers is not in the document]
Comment (xross) [reply tests/test_bulk_rx_traffic.py]: Done

File: tests/test_bulk_tx_badack.py [hunk 8]
```diff
@@ -1,53 +1,31 @@
...
-    packets = []
-
-    dataval = 0;
-
-    pid = PID_DATA0;
-
-    pktlength = 10
+    session = UsbSession(bus_speed = usb_speed, run_enumeration = False, device_address = address)
+    
+    for pktLength in range(10, 14):
```
[the remark this answers is not in the document]
Comment (xross) [reply tests/test_bulk_tx_badack.py:19]: We need to test a good range of values ideally. These were chosen as there are separate code paths different "tail" lengths - word aligned packet length, word aligned + 1 byte,  word aligned +2 bytes and word aligned + 3 bytes

Also, short packets tend to be harder to meeting timing

File: tests/usb_packet.py (hunk 1 above)
Comment (xross) [reply tests/usb_packet.py:347]: yup. Done.

File: tests/usb_packet.py (hunk 2 above)
Comment (xross) [reply tests/usb_packet.py]: Done

File: tests/usb_packet.py (hunk 3 above)
Comment (xross) [reply tests/usb_packet.py:489]: Added a TODO comment for XS2 and removed commented out code

File: tests/usb_transaction.py (hunk 4 above)
Comment (xross) [reply tests/usb_transaction.py:75]: Removed

**Review by DaveAtkinsonXMOS — APPROVED** [review]
Looks good to me.

**Pull request merged**